### PR TITLE
3197 DB schema: change_properties to text

### PR DESCRIPTION
**Pull request by nand0p**
Alters change_properties.propery_value from String(1024) to Text

http://trac.buildbot.net/ticket/3197

### CODE REVIEW

**Comment by nand0p** [comment]
The upgrade is working, but the test is not. i will keep working thru this.

before:
`
==> buildbot: CREATE TABLE change_properties (
==> buildbot: 	changeid INTEGER NOT NULL, 
==> buildbot: 	property_name VARCHAR(256) NOT NULL, 
==> buildbot: 	property_value VARCHAR(1024) NOT NULL, 
==> buildbot: 	FOREIGN KEY(changeid) REFERENCES changes (changeid)
==> buildbot: );
`

after:
`
==> buildbot: CREATE TABLE change_properties (
==> buildbot: 	changeid INTEGER NOT NULL, 
==> buildbot: 	property_name VARCHAR(256) NOT NULL, 
==> buildbot: 	property_value TEXT NOT NULL, 
==> buildbot: 	FOREIGN KEY(changeid) REFERENCES changes (changeid)
==> buildbot: );
`

**Comment by nand0p** [comment]
@tardyp It looks like this issue may only be occuring with postgresql9.4 and latest sqlalchemy, as the CI tests show.  

After silently executing, the data_type remains unchanged:

> ==> buildbot: 2017-01-09 20:19:33+0000 [-] Log opened.
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] checking basedir
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] checking for running master
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] checking master.cfg
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] Loading configuration from '/tmp/bb-mig-1483993162/master.cfg'
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] upgrading basedir
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] creating /tmp/bb-mig-1483993162/master.cfg.sample
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] upgrading database (postgresql://buildbot:xxxx@localhost/bbdb)
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] Setting up database with URL 'postgresql://buildbot:xxxx@localhost/bbdb'
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] using SQLAlchemy-Migrate version 0.10.0
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] migrating schema version 47 -> 48
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] upgrade complete
> ==> buildbot: 2017-01-09 20:19:33+0000 [-] Main loop terminated.
> ==> buildbot: test db schema:
> ==> buildbot:   column_name   |     data_type     | character_maximum_length 
> ==> buildbot: ----------------+-------------------+--------------------------
> ==> buildbot:  changeid       | integer           |                         
> ==> buildbot:  property_name  | character varying |                      256
> ==> buildbot:  property_value | character varying |                     1024

**Comment by tardyp** [comment]
@nand0p indeed, this is the pain of the db migrations it never works easily for the 3 dbs we support. That's why we require the CI to pass for those 3 :-}

There is already a change property to text migration 

https://github.com/buildbot/buildbot/blob/eight/master/buildbot/db/migrate/versions/023_increase_comments_property_lengths.py

File: master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    def test_update(self):
+        def setup_thd(conn):
+            self.create_table_thd(conn)
+
+        def verify_thd(conn):
+            metadata = sa.MetaData()
+            metadata.bind = conn
+
+            change_properties = sautils.Table('change_properties', metadata, autoload=True)
+            self.assertIsInstance(change_properties.c.property_value.type, sa.Text)
```
Comment (tardyp) [inline master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py]: I think for sqlite, we dont really care as TEXT and VARCHAR are stored the same anyway. We rather need to check that the property can store > 1024 string.

Comment (nand0p) [inline master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py]: makes sense.  i will add this test.

File: master/buildbot/newsfragments/db_schema_change_properties.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Alters change properties from String to Text. Requires upgrade master. (:bug:`3197`)
```
Comment (tardyp) [inline master/buildbot/newsfragments/db_schema_change_properties.bugfix]: Please change to more user oriented message:

    Change properties can now be of unlimited size (previously was max 1024 bytes)

File: master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py (hunk 1 above)
Comment (tardyp) [inline master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py]: I dont think you added that test

**Comment by nand0p** [comment]
I can't seem to replicate the mysql issue seen in CI.  @tardyp should I ignore the insert test for mysql instead?

> ==> buildbot: 2017-01-11 20:21:27+0000 [-] Log opened.
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] checking basedir
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] checking for running master
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] checking master.cfg
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] Loading configuration from '/tmp/bb-mig-1484166077/master.cfg'
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] upgrading basedir
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] creating /tmp/bb-mig-1484166077/master.cfg.sample
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] upgrading database (mysql://buildbot:xxxx@localhost/bbdb)
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] Setting up database with URL 'mysql://buildbot:xxxx@localhost/bbdb'
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] using SQLAlchemy-Migrate version 0.10.0
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] migrating schema version 47 -> 48
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] upgrade complete
> ==> buildbot: 2017-01-11 20:21:27+0000 [-] Main loop terminated.
> ==> buildbot: test db schema:
> ==> buildbot: Field	Type	Null	Key	Default	Extra
> ==> buildbot: changeid	int(11)	NO	MUL	NULL	
> ==> buildbot: property_name	varchar(256)	NO		NULL	
> ==> buildbot: property_value	text	NO		NULL	
> ==> buildbot: mysqld  Ver 5.7.17 for Linux on x86_64 (MySQL Community Server (GPL))
> ==> buildbot: run tests:
> ==> buildbot: # On branch 3197_change_properties_to_text
> ==> buildbot: nothing to commit, working directory clean
> ==> buildbot: buildbot.test.unit.test_db_migrate_versions_048_change_properties_to_text
> ==> buildbot:     test_update ...                                                        [OK]

**Comment by nand0p** [comment]
@tardyp ready_for_review

File: master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    def test_update(self):
+        def setup_thd(conn):
+            self.create_table_thd(conn)
+
+        def verify_thd(conn):
+            metadata = sa.MetaData()
+            metadata.bind = conn
+            random_length = 4096
+            random_string = ''.join(choice(ascii_lowercase) for byte in range(random_length))
```
Comment (tardyp) [inline master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py]: well "x"*4096 would have been enough..

File: master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                dict(changeid=1,
+                     property_name="test_change_properties_property_value_length",
+                     property_value=random_string),
+            ])
+
+            # Test db string length
+            # Using raw sql due to error via sa
+            # q = conn.execute(sa.select([sa.func.length(change_properties.c.property_value)]).where(changeid==1))
+            q = conn.execute("select length(property_value) from change_properties where changeid = 1")
+            [self.assertEqual(q_length[0], random_length) for q_length in q]
```
Comment (tardyp) [inline master/buildbot/test/unit/test_db_migrate_versions_048_change_properties_to_text.py]: I would have prefered you just use the db api to write and read the property back and forth and make sure its not truncated

**Comment by nand0p** [comment]
@tardyp You're right, there does appear to be an issue with sqlalchemy truncating text fields when they contain large strings. To compensate, I kept the type as string(varchar), but upped to 14336 (1024*14), to deal with mysql limits. http://stackoverflow.com/questions/13506832/what-is-the-mysql-varchar-max-size

should we worry about truncation for any of these?

> ./master/buildbot/db/model.py:        sa.Column('value', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('state_string', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('state_string', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('urls_json', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('name', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('property_value', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('parent_relationship', sa.Text),
> ./master/buildbot/db/model.py:        sa.Column('name', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('comments', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('patch_base64', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('patch_author', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('patch_comment', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('subdir', sa.Text),
> ./master/buildbot/db/model.py:        sa.Column('name', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('name', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('description', sa.Text, nullable=True),
> ./master/buildbot/db/model.py:        sa.Column('name', sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column("value_json", sa.Text, nullable=False),
> ./master/buildbot/db/model.py:        sa.Column('name', sa.Text, nullable=False),

File: master/buildbot/db/model.py [hunk 5]
```diff
@@ -314,7 +314,7 @@ class Model(base.DBConnectorComponent):
                   nullable=False),
         sa.Column('property_name', sa.String(256), nullable=False),
         # JSON-encoded tuple of (value, source)
-        sa.Column('property_value', sa.String(1024), nullable=False),
+        sa.Column('property_value', sa.String(14336), nullable=False),
```
Comment (tardyp) [inline master/buildbot/db/model.py]: I dont think it is a good idea. We should keep Text, and be coherent with the other property table.

**Comment by nand0p** [comment]
@tardyp ready_for_review

**Comment by tardyp** [comment]
I dont see the documentation about mysql limitation.

Can you document it here (with reference on mysql documentation on why it truncates the text rows)
https://github.com/tardyp/buildbot/blob/master/master/docs/manual/cfg-global.rst#L66

**Comment by nand0p** [comment]
The mysql limitation only applies to varchar, not text, so it is not relevant anymore now that the table column is altered to text.  

The truncation I was seeing apparently was a result of evaluating the raw unicode result string. by encoding as "ascii" before testing, this issue is resolved.

example raw unicode string:

> (u'szejzjapotmsjlmyeeixifhpnrwwavtqkvycywhaujatezhrfgzzoiuwocbfzyqydxcgnmcbpczhoabhqkwejwfksgqondbuihfeqjslrixrshhgsgxunfjticykltbuguxwtegxxfbodtusabqe ... (65238 characters truncated) ... wcppgackzpxvasrfjdiazbxhvtjzqggzlueeoiwntoeiovvhorgbtvuhneojtkkhvioadgcwrozkfqairporuxmybewweigaapzyhlhcuthearlhwotqpzidulftveztyuhldahokuhaxrhjxysaq',)

**Comment by tardyp** [comment]
thanks!

**Pull request merged**